### ISSUE TITLE
Backend registry

### ISSUE THREAD

**Issue by SylvainCorlay** [jupyter-widgets/ipywidgets#1756]
Reporting the conversation on gitter

In ipywidgets 7.0, the unique backend key for the widget registry has been replaced with the concatenation of all the items of the model spec

(model module, model name, model module version, view module, view name, view module version).

https://github.com/jupyter-widgets/ipywidgets/blob/master/ipywidgets/widgets/widget.py#L263

I am confused with the reason for

 - including the version.
 - the commented code about semver-matching the model and view version.

since I don't see how more than one version of the combination model/view - module/name can exist.

Incidentally, the use of a backend key like in ipywidgets 7.0 would have made the xwidgets implementation of the protocol easier. The reason is that the backend key could have been a static property of the widget type, while the spec is a runtime property, as it is part of the widget state.

**Comment by jasongrout** [comment]
> I cannot think of a situation where more than one version of the combination model/view - module/name can exist.

I must be missing something in what you are saying. As an example, in the upgrade to ipywidgets 7, many model/view combinations in third-party widgets had a version number upgrade, without changing the names or modules.


> Incidentally, the use of a backend key like in ipywidgets 6.0 would have made the xwidgets implementation of the protocol easier.

That 'backend key' (or I would say 'model spec name') then requires another version number to deal with. Personally, I would like to have a single spec and version number for that spec. I thought you were against introducing new version numbers other than the existing js package versions.

Sounds like we should have a conversation about this (which I know you've been trying to do - sorry I've been so busy!)

**Comment by SylvainCorlay** [comment]
> > I cannot think of a situation where more than one version of the combination model/view - module/name can exist.
>
> I must be missing something in what you are saying. As an example, in the upgrade to ipywidgets 7, many model/view combinations in third-party widgets had a version number upgrade, without changing the names or modules.

Yes, what I meant, is that in a given python package (installed in the backend), there cannot be two widgets having the same model/view names and modules, but differing in they version requirements, unless you want to introduce some separate classes to handle deprecated "model spec"...

**Comment by SylvainCorlay** [comment]
In `xwidgets`, the `xfactory` is responsible for instantiating widgets upon reception of comm open messages.

https://github.com/QuantStack/xwidgets/blob/master/include/xwidgets/xfactory.hpp

At the moment, I am merely concatenating `model_module + model_name + view_module + view_name`
